### PR TITLE
BHBC-1364 + BHBC-1365

### PR DESCRIPTION
**Pull request by NickPhura**
# Overview

## Links to jira tickets

https://quartech.atlassian.net/browse/BHBC-1328

## This PR contains the following changes

Initial transformations endpoint + code to translate an Excel template into a DWC Archive.
- Supports pivoting occurrence counts
- Supports relationships between records (for fields like cow w/1 calf)

## This PR contains the following types of changes

- [ ] New feature (change which adds functionality)
- [ ] Enhancement (improvements to existing functionality)
- [ ] Bug fix (change which fixes an issue)
- [ ] Misc cleanup / Refactoring / Documentation
- [ ] Version change

## Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### General

- [ ] I have performed a self-review of my own code

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [ ] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [ ] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [ ] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Screenshots

Please add any relevant UI screenshots, if applicable.

### CODE REVIEW

**Comment by NickPhura** [comment]
Test file I was using: [Moose_SRB_Composition_Survey_2.0.xlsx](https://github.com/bcgov/biohubbc/files/7282126/Moose_SRB_Composition_Survey_2.0.xlsx)

File: package-lock.json
Comment (NickPhura) [inline package-lock.json:1]: This was a stray lock file at the root level, probably added accidentally sometime.

File: Makefile
Comment (sdevalapurkar) [inline Makefile]:
```suggestion
cypress: ## Runs `npm test:e2e` for api, and app projects
```

File: api/src/database/db.test.ts
Comment (sdevalapurkar) [inline api/src/database/db.test.ts]: why do we need this in this file?

File: api/src/paths/xlsx/transform.ts
Comment (sdevalapurkar) [inline api/src/paths/xlsx/transform.ts]: this commented code being left here is intentional right?

File: api/src/paths/xlsx/transform.ts
Comment (sdevalapurkar) [inline api/src/paths/xlsx/transform.ts]: could you add a comment with an example?
```
/project/1/blah.txt -> /project/1
```

File: api/src/paths/xlsx/transform.ts
Comment (sdevalapurkar) [inline api/src/paths/xlsx/transform.ts]: can this be defined once somewhere and just exported and used amongst various endpoints?

File: api/src/utils/media/xlsx/transformation/transformation-schema-parser.ts
Comment (sdevalapurkar) [inline api/src/utils/media/xlsx/transformation/transformation-schema-parser.ts]: nit:
```suggestion
      throw Error('TransformationSchemaParser - provided validationSchema was not valid JSON');
```

File: app/src/features/surveys/view/SurveyObservations.tsx
Comment (sdevalapurkar) [inline app/src/features/surveys/view/SurveyObservations.tsx]: remove this?

File: database/src/migrations/template_methodology_species_transformations/moose_recruitment_survey.json
Comment (sdevalapurkar) [inline database/src/migrations/template_methodology_species_transformations/moose_recruitment_survey.json:1]: question:
is this supposed to be empty?

File: api/src/database/db.test.ts
Comment (NickPhura) [reply api/src/database/db.test.ts]: It was complaining before, though I dont remember why exactly.  I think it was claiming it couldn't find some functions.

File: api/src/paths/xlsx/transform.ts
Comment (NickPhura) [reply api/src/paths/xlsx/transform.ts]: Only partially.  I am not 100% certain what the workflow should be if transformation fails, but validation passes, etc.

File: api/src/paths/xlsx/transform.ts
Comment (NickPhura) [reply api/src/paths/xlsx/transform.ts]: Yes, but I was holding off for now.  I want to move all of our re-useable request handlers into some other folder.  Primarily because you run into circular-dependencies very quickly if all of our endpoints are importing from one-another.

Comment (NickPhura) [reply api/src/paths/xlsx/transform.ts]: Ill try just keeping the version in dwc/validate works, for now.

File: api/src/database/db.test.ts
Comment (sdevalapurkar) [reply api/src/database/db.test.ts]: that's weird... not too sure why that would be the case 😕

File: database/src/migrations/template_methodology_species_transformations/moose_recruitment_survey.json
Comment (NickPhura) [reply database/src/migrations/template_methodology_species_transformations/moose_recruitment_survey.json:1]: yeah, I just hadn't looked into it yet. Could do that before this gets merged tho, or I can remove it from the current migration and add it after.  Either way.

File: app/src/features/surveys/view/SurveyObservations.tsx
Comment (NickPhura) [reply app/src/features/surveys/view/SurveyObservations.tsx]: removed

File: api/src/utils/media/xlsx/transformation/transformation-schema-parser.ts
Comment (NickPhura) [reply api/src/utils/media/xlsx/transformation/transformation-schema-parser.ts]: fixed

File: api/src/paths/xlsx/transform.ts
Comment (NickPhura) [reply api/src/paths/xlsx/transform.ts]: Fixed

File: api/src/paths/xlsx/transform.ts
Comment (NickPhura) [reply api/src/paths/xlsx/transform.ts]: Added

File: database/src/migrations/template_methodology_species_transformations/moose_recruitment_survey.json
Comment (sdevalapurkar) [reply database/src/migrations/template_methodology_species_transformations/moose_recruitment_survey.json:1]: whichever you think is better :)

Comment (NickPhura) [reply database/src/migrations/template_methodology_species_transformations/moose_recruitment_survey.json:1]: Ill see if I can get it in today, in this PR.  Probably the least amount of work that way

**Comment by anissa-agahchen** [comment]
It looks good.  I tested quite a few scenarios .. and the transformation works really well.  Couple of things, and this might be outside the scope of this first pass .... how to we capture that an input file has no data to be transformed? also when template validation fails.  For example, I put in 'a' for a lone cow, the validation failed, the transform still went through and you see the 'a' in the occurrence file
Screen Shot 2021-10-05 at 1 27 06 PM

**Comment by NickPhura** [comment]
@anissa-agahchen 
>It looks good. I tested quite a few scenarios .. and the transformation works really well. Couple of things, and this might be outside the scope of this first pass .... how to we capture that an input file has no data to be transformed? 

Ill need to account for that, Ill look into it.

> also when template validation fails. For example, I put in 'a' for a lone cow, the validation failed, the transform still went through and you see the 'a' in the occurrence file

Right, this is something I can adjust in the n8n workflow.  Check if the status of the request was successful, and if not dont run the next steps.

File: Makefile
Comment (NickPhura) [reply Makefile]: fixed

**Comment by anissa-agahchen** [comment]
@NickPhura ... just the geometry piece left now on the occurrencefeaturegroup.tsx when I switch the methodology from SRB/Composition to Recruitment.

File: api/src/database/db.test.ts
Comment (NickPhura) [reply api/src/database/db.test.ts]: I took it out and it seemed fine now.  So no idea what I had seen before :/

File: database/src/migrations/template_methodology_species_transformations/moose_recruitment_survey.json
Comment (NickPhura) [reply database/src/migrations/template_methodology_species_transformations/moose_recruitment_survey.json:1]: This will need robins input, and probably will take her a bit.  So will resolve in another PR later

**Review by anissa-agahchen — DISMISSED** [review]

**Review by anissa-agahchen — APPROVED** [review]

**Pull request merged**
